### PR TITLE
feat(delete-user-data): add auto discovery mechanisms + update documentation 

### PR DESCRIPTION
**Pull request by dackers86**
Introducing new options for searching and deleting user data automatically once a user has been deleted from Firebase Authentication.

Includes: 

- [x] Firestore Collection Queries: configurable queries which are executed when a user has been deleted.
- [x] Firestore Search: Automatically searches and removes user data from any root level collections, document and documents based on ids and user specified fields. 
- [x] Includes a function parameters for allowing users to specify there own field paths to be deleted.
- [x] Increased test coverage to ensure new functionality works, includes compatibility with existing delete user data functionality.
- [x] Uses pub-sub to scale and enhance performance.

### CODE REVIEW

**Comment by huangjeff5** [comment]
For configurable queries I think it makes more sense to expose a hook, versus using Firestore.

Hooks are at least as powerful as defining a custom syntax for collection queries in Firestore. You could do things that you might not be able to express in a query. 

In terms of cognitive overhead, since we have a custom data structure they need to use to express the query, I feel like there's also more to learn versus a hook function.

File: delete-user-data/extension.yaml [hunk 1]
```diff
@@ -213,3 +231,28 @@ params:
     type: string
     example: "{DEFAULT}/{UID}-pic.png,my-awesome-app-logs/{UID}-logs.txt"
     required: false
+
+  - param: FIRESTORE_QUERY_COLLECTION
```
Comment (huangjeff5) [inline delete-user-data/extension.yaml:235]: Isn't this param a duplicate of line 152?

File: delete-user-data/functions/src/helpers.ts [hunk 2]
```diff
@@ -9,3 +13,195 @@ export const getDatabaseUrl = (
...
+  await batch.commit();
+
+  // Recurse on the next process tick, to avoid
+  // exploding the stack.
+  process.nextTick(() => {
+    deleteQueryBatch(db, query, resolve);
+  });
+}
+
+export async function repeat(
```
Comment (huangjeff5) [inline delete-user-data/functions/src/helpers.ts:50]: Is this used?

File: delete-user-data/functions/src/helpers.ts [hunk 3]
```diff
@@ -9,3 +13,195 @@ export const getDatabaseUrl = (
...
+        if (!timedOut) {
+          clearTimeout(timer);
+          resolve(snapshot);
+        }
+      }
+    });
+  });
+};
+
+export const waitForDocumentToExistInCollection = (
```
Comment (huangjeff5) [inline delete-user-data/functions/src/helpers.ts:116]: Is this used?

File: delete-user-data/functions/src/helpers.ts [hunk 4]
```diff
@@ -9,3 +13,195 @@ export const getDatabaseUrl = (
...
+        if (!timedOut) {
+          clearTimeout(timer);
+          resolve(snapshot);
+        }
+      }
+    });
+  });
+};
+
+export const waitForDocumentUpdate = (
```
Comment (huangjeff5) [inline delete-user-data/functions/src/helpers.ts:92]: Is this used?

File: delete-user-data/functions/src/helpers.ts [hunk 5]
```diff
@@ -9,3 +13,195 @@ export const getDatabaseUrl = (
...
+      await new Promise((r) => setTimeout(r, interval));
+      return repeat(fn, until, retriesLeft - 1, interval);
+    }
+    throw new Error("Max repeats count reached");
+  }
+
+  return result;
+}
+
+export const waitForDocumentToExistWithField = (
```
Comment (huangjeff5) [inline delete-user-data/functions/src/helpers.ts:69]: is this used?

File: delete-user-data/functions/src/index.ts [hunk 6]
```diff
@@ -36,6 +39,56 @@ admin.initializeApp({
...
+      if (doc.path.includes(uid)) {
+        await doc.delete();
+      }
+
+      /** Check field/value */
+      const data = await doc.get();
+      const fields = data.data();
+      if (fields) {
+        Object.entries(fields).forEach(async ([field, value]) => {
+          if (field.includes(uid)) {
```
Comment (huangjeff5) [inline delete-user-data/functions/src/index.ts:71]: Is this logic correct? You're deleting it if the document contains the Auth UID as a fieldname in the document?

I was thinking the user has to specify a auth_user_id_fieldname parameter (defaults to "userId"). The value of that is what should be matching Auth UID.

File: delete-user-data/functions/src/helpers.ts (hunk 5 above)
Comment (dackers86) [reply delete-user-data/functions/src/helpers.ts:69]: It doesn't seem to be required. Removed.

File: delete-user-data/functions/src/helpers.ts (hunk 4 above)
Comment (dackers86) [reply delete-user-data/functions/src/helpers.ts:92]: Also removed.

File: delete-user-data/functions/src/helpers.ts (hunk 2 above)
Comment (dackers86) [reply delete-user-data/functions/src/helpers.ts]: Removed.

File: delete-user-data/functions/src/helpers.ts (hunk 3 above)
Comment (dackers86) [reply delete-user-data/functions/src/helpers.ts]: Removed.

File: delete-user-data/functions/src/index.ts (hunk 6 above)
Comment (dackers86) [reply delete-user-data/functions/src/index.ts]: Refactored with new query lookup.

File: delete-user-data/extension.yaml [hunk 7]
```diff
@@ -49,6 +49,8 @@ roles:
     reason: Allows the extension to delete (user) data from Realtime Database.
   - role: storage.admin
     reason: Allows the extension to delete (user) data from Cloud Storage.
+  - role: pubsub.admin
+    reason: allows publish and subscribing of deletion events
```
Comment (Ehesp) [inline delete-user-data/extension.yaml:53]:
```suggestion
    reason: Allows the extension to publish and subscribe to PubSub deletion events.
```

File: delete-user-data/extension.yaml [hunk 8]
```diff
@@ -193,6 +211,17 @@ params:
     example: users/{UID},admins/{UID}
     required: false
 
+  - param: STORAGE_BUCKET
+    label: Cloud Storage bucket
+    description: >
+       Where in Google Cloud Storage do you store user data?
```
Comment (Ehesp) [inline delete-user-data/extension.yaml:217]: This description is incorrect.

File: delete-user-data/extension.yaml [hunk 9]
```diff
@@ -212,4 +241,76 @@ params:
       `{DEFAULT}/{UID}-pic.png,my-app-logs/{UID}-logs.txt,{DEFAULT}/media/{UID}`.
     type: string
     example: "{DEFAULT}/{UID}-pic.png,my-awesome-app-logs/{UID}-logs.txt"
+    required: true
+
+  - param: ENABLE_AUTO_DISCOVERY
```
Comment (Ehesp) [inline delete-user-data/extension.yaml:246]: Requires updated validation regex.

File: delete-user-data/extension.yaml [hunk 10]
```diff
@@ -212,4 +241,76 @@ params:
...
+  - type: firebase.extensions.delete-user-data.v1.firestore
+    description: Occurs when a set of Firestore paths has been marked for deletion following a delete user trigger.
+
+  - type: firebase.extensions.delete-user-data.v1.rtdb
+    description:  Occurs when a set of Realtime Database paths has been marked for deletion following a delete user trigger.
+
+  - type: firebase.extensions.delete-user-data.v1.storage
+    description:  Occurs when a set of Storage paths has been marked for deletion following a delete user trigger.
+
+  - type: firebase.extensions.delete-user-data.v1.autoDiscovery
```
Comment (Ehesp) [inline delete-user-data/extension.yaml:314]: This event should be removed.

File: delete-user-data/extension.yaml [hunk 11]
```diff
@@ -65,6 +67,22 @@ resources:
...
+
+  - name: handleDeletion
+    type: firebaseextensions.v1beta.function
+    properties:
+      location: ${LOCATION}
+      eventTrigger:
+        eventType: google.pubsub.topic.publish
+        resource: projects/${PROJECT_ID}/topics/${param:DELETION_TOPIC}
+
 params:
```
[the remark this answers is not in the document]
Comment (dackers86) [reply delete-user-data/extension.yaml:86]: These are not referenced in the docs. Should they be?

**Review by cabljac — APPROVED** [review]
lgtm!

**Pull request merged**